[RFC DO NOT MERGE]SOF D0ix support

keyonjie opened the pull request:
This series is to enable SOF D0ix support from driver side. What it does including:

1. implement helpers to update d0i3c register for d0i0<->d0i3.
2. add fields to store counts of active PCM and active low power PCM, increase/decrease them at PCM open/close.
3. add field to store dsp_state and update it when needed.
4. add a delay work for d0ix transition, schedule the work when needed.

Todos:
1. destroy/restore unused pipelines/widgets at d0ix transition. please refer to: thesofproject/linux#1146
2. PM ops support. For suspend/resume, how should we handle d0ix, is still under discussion, s0ix is not supported yet, will update soon.
3. IPC over register is not included in this PR.

keyonjie commented:
The corresponding FW update in:
https://github.com/thesofproject/sof/pull/1748

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -65,6 +65,10 @@
 
+/* D0I3C Register fields */
+#define AZX_REG_VS_D0I3C_CIP		BIT(0) /* Command-In-Progress */
+#define AZX_REG_VS_D0I3C_I3		BIT(2) /* D0i3 enable bit */
```
if they are AZX registers they should not be in an SOF file?
Is this redefined from somewhere else?

keyonjie replied:
they are bits of vendor specific AZX register AZX_REG_VS_D0I3C:
```
include/sound/hda_register.h:123:#define AZX_REG_VS_D0I3C		0x104A
```
And bits definition is not included in include/sound/hda_register.h, SST driver define them in skl.h, do you think we should rename them to SOF_HDA_VS_D0I3C_xxx?

plbossart replied:
if those bits are only defined in SOF add the SOF prefix. If those bits exist somewhere else don't redefine them.
We only care about legacy/SOF, what SKL/SST folks did is not something we should blindly follow.

keyonjie replied:
OK, let me rename them to be SOF_HDA_VS_D0I3C_xxx.

plbossart commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -60,6 +60,9 @@
 #define SOF_TKN_PCM_DMAC_CONFIG			353
 
+/* Stream */
+#define SOF_TKN_STREAM_AVAIL_ON_LP		380
```
LP is too vague.

and it's not that the stream is available, it's that it remains enabled during D0ix transitions even if there is no data transferred.

Need to find a different wording.

Also you'd want to start a new category for power-related tokens, aligned on a 100 boundary.

keyonjie replied:
When went through topology PCM definition today, I realized that we don't need introduce new token for this, I will change to reuse the existed voice_wakeup flag in the next update.

plbossart replied:
no. this D0ix can be used for playback as well, e.g with 2s ring buffers. Don't overload a processing capability with a power capability.

keyonjie replied:
OK, so w.r.t. the wording, active_on_lp or suspend_active are both confusing to me, it can be active on low-power(or on suspended, or in D0i3), it can be stopped (on low-power/suspend/D0i3) also? I wanted to illustrate the capability of the stream, that's why I used the wording "available", it can be used on low-power, but you can stop it at suspend also.

@plbossart what do you think?

keyonjie replied:
> Need to find a different wording.
> 
> Also you'd want to start a new category for power-related tokens, aligned on a 100 boundary.

I thought this is stream capability, that's why I put it here near the "PCM" category, and there is no 100 boundary left here. Are you proposing to put it to the end, e.g.
#define SOF_TKN_STREAM_AVAIL_ON_LP		1000

plbossart replied:
#define SOF_TKN_STREAM_COMPATIBLE_D0ix	1000 ?

keyonjie replied:
> #define SOF_TKN_STREAM_COMPATIBLE_D0ix 1000 ?

sounds good to me, let me change.

keyonjie replied:
@plbossart as stated below, since we are aligned to set the flag direction dependent, as 1000 and 1100 has been occupied(in driver side only atm, not updated/aligned in topology), so I want to add it like this:
```
diff --git a/tools/topology/sof/tokens.m4 b/tools/topology/sof/tokens.m4
index 93645ceb..d01fa55d 100644
--- a/tools/topology/sof/tokens.m4
+++ b/tools/topology/sof/tokens.m4
@@ -42,6 +42,8 @@ SectionVendorTokens."sof_src_tokens" {
 }
 
+ SectionVendorTokens."sof_stream_tokens" {
+       SOF_TKN_STREAM_PLAYBACK_COMPATIBLE_D0I3            "1200"
+       SOF_TKN_STREAM_CAPTURE_COMPATIBLE_D0I3             "1201"
+ }
```
@plbossart @ranj063 sounds good?

plbossart replied:
ok for me, @ranj063 you good?

ranj063 replied:
@plbossart @keyonjie yes sounds good

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -724,6 +724,13 @@ static const struct sof_topology_token pcm_tokens[] = {
+/* PCM */
+static const struct sof_topology_token stream_tokens[] = {
+	{SOF_TKN_STREAM_AVAIL_ON_LP,
+		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
```
isn't this a BOOL? why 32 bits?

keyonjie replied:
will be removed.

keyonjie replied:
the BOOL type doesn't work, but no worry, let me submit a separate PR to fix it, and then change this use BOOL.

keyonjie replied:
That is https://github.com/thesofproject/linux/pull/1165, @plbossart

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -293,6 +293,7 @@ struct snd_sof_pcm_stream {
 	struct sof_ipc_stream_posn posn;
 	struct snd_pcm_substream *substream;
 	struct work_struct period_elapsed_work;
+	u32 avail_at_lp; /* pcm can run at low power like d0i3 */
```
again bad name.

suspend_active or something like this would make more sense.

keyonjie replied:
will be remove, and reuse the dai_link->ignore_suspend.

plbossart replied:
no. ignore_suspend is a hack. I'll have to dig in my archives but it doesn't cover all cases.

keyonjie replied:
OK, then let me give up that and keep using vendor token for it.

ranj063 replied:
@keyonjie please squash with previous commit. otherwise you'll break git bisect with compilation errors with the previous commit.

keyonjie replied:
> @keyonjie please squash with previous commit. otherwise you'll break git bisect with compilation errors with the previous commit.

indeed, thanks. I previously planed to put this commit before the previous one.

keyonjie replied:
@ranj063 actually this commit is before the token adding and token parsing ones in my personal branch:
https://github.com/keyonjie/linux/commits/pm_d0ix

don't know why the sequence is changed in the PR submitting, :-D

```
ASoC: SOF: topology: parse and store avail_at_lp flag …

@keyonjie
keyonjie committed 27 days ago
 
ASoC: SOF: token: add token for low power PCM. …

@keyonjie
keyonjie committed 27 days ago
 
ASoC: SOF: add flag to snd_sof_pcm_stream for low power stream …

@keyonjie
keyonjie committed 27 days ago
 
ASoC: SOF: Intel: hda-dsp: Add helper for setting dsp to d0i3 …

@keyonjie
keyonjie committed 28 days ago
```

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2329,6 +2339,15 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
+			le32_to_cpu(private->size));
+		return ret;
+	}
+
```
question: do we see a case where we have D0ix support for playback and capture simultaneously? Or is this possible but not really necessary?

keyonjie replied:
I think it is possible but not really necessary, what I can imagine at the moment is that only KWD capture can run at D0ix, but yes, theoretically, we should make it possible for playback(or even playback/capture simultaneously) at D0ix.

ranj063 replied:
@keyonjie @plbossart we seem to use the same token for both playback and capture here. Is it possible that only one of the directions is D0ix compatible and not the other?

plbossart replied:
yes in theory, you could have an imbalance and only have large buffers for playback and not capture. So this should be direction dependent indeed.

keyonjie replied:
yes, and, if we want to make it direction dependent, we might need to define 2 different tokens, one for playback, one for capture, as the token is located in the data segment of the SectionPCM.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -371,6 +371,10 @@ struct snd_sof_dev {
 	struct snd_soc_component_driver plat_drv;
 
+	/* power status related */
+	u32 pcm_active;		/* Count of active PCMs */
```
why is this needed? This is already handled by pm_runtime.

keyonjie replied:
um, good point, let me think it more.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -371,6 +371,10 @@ struct snd_sof_dev {
 
+	/* power status related */
+	u32 pcm_active;		/* Count of active PCMs */
+	u32 lp_pcm_active;	/* Count of active low power PCMs */
```
if you have this capability for playback and capture, does this increase the refcount by 1 or 2?

keyonjie replied:
the refcount will increase by 1 at each substream open(), so if you open both playback and capture, the refcount will increase by 2.

plbossart replied:
then make it clear that it's for each substream and direction in the comments.

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -486,10 +486,17 @@ static int sof_pcm_open(struct snd_pcm_substream *substream)
+	/* update active pcm counts */
+	sdev->pcm_active++;
+	if (spcm->stream[substream->stream].avail_at_lp)
+		sdev->lp_pcm_active++;
```
don't we need a mutex here?

keyonjie replied:
We don't need mutex or lock as it is used in open/close which is already managed by ALSA framework.

plbossart replied:
then add a comment to show that you know what you are doing :-)

keyonjie replied:
OK, will add it.

plbossart replied:
if you say you will add a comment then do it...

keyonjie replied:
oops.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -522,6 +529,11 @@ static int sof_pcm_close(struct snd_pcm_substream *substream)
+	/* update active pcm counts */
+	sdev->pcm_active--;
+	if (spcm->stream[substream->stream].avail_at_lp)
+		sdev->lp_pcm_active--;
```
and here? locking seems required for refcounting?

keyonjie replied:
Same as at .open(), we don't need mutex or locking.

plbossart replied:
same, please take into account reviews...

keyonjie replied:
Sorry about missing this. Looks the series is becoming too huge for me.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -357,6 +357,13 @@ struct snd_sof_dai {
+enum sof_dsp_state {
+	SOF_DSP_D3, /* DSP in reset state */
+	SOF_DSP_D0I3, /* DSP Running in D0i3(low power) state*/
+	SOF_DSP_D0I0, /* DSP Running in D0 state */
```
why is this needed?
D3/D0I0 are already pm_runtime states. I don't know how you'd synchronize the value of this enum with the pm_runtime core.

keyonjie replied:
I added D3 and D0I0 to make the states integrated and easier for transition in state machine. Let me think about how to combine it with the system device PM states.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -385,6 +387,8 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 			"error: failed to power down DSP during suspend %d\n",
 			ret);
 
+	sdev->dsp_state = SOF_DSP_D3;
```
it's really unclear what this helps with?

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -193,6 +193,15 @@ static inline int snd_sof_dsp_set_clk(struct snd_sof_dev *sdev, u32 freq)
 }
 
+static inline int snd_sof_dsp_d0ix_update(struct snd_sof_dev *sdev,
+					  bool d0i3_enable)
```
that doesn't seem to align with @lbetlej initial proposal. we wanted to have an IPC where you can set the state, not just a boolean.

keyonjie replied:
are you talking about PM_STATE_SET IPC between SW and FW? I haven't implemented IPC part yet(in the Todo list), please see inline:

```
	if (sdev->dsp_state == SOF_DSP_D0I0 && d0i3_enable) {
		dev_dbg(sdev->dev, "debug: entering to D0I3...\n");

		/* Todo: destroy unused pipelines/widgets in FW */

		/* do platform specific d0i3 entering task */
		ret = snd_sof_dsp_d0ix_update(sdev, d0i3_enable);

		/* sending d0i3_entry IPC */ //Pierre: I will add IPC sending PM_STATE_SET with D0I3 state here

		/* update dsp state */
		sdev->dsp_state = SOF_DSP_D0I3;

		return ret;
	}

	if (sdev->dsp_state == SOF_DSP_D0I3 && !d0i3_enable) {
		dev_dbg(sdev->dev, "debug: exiting from D0I3...\n");

		/* Todo: reload unused pipelines/widgets in FW */

		/* do platform specific d0i3 entering task */
		ret = snd_sof_dsp_d0ix_update(sdev, d0i3_enable);

		/* sending d0i3_exit IPC */ //Pierre: I will add IPC sending PM_STATE_SET with D0I0 state here

		/* update dsp state */
		sdev->dsp_state = SOF_DSP_D0I0;

		return ret;
	}
```

plbossart replied:
if you are planning to send a state, then make it visible at a higher level instead of a boolean that will result in a hard-coded decision.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -423,3 +423,86 @@ int snd_sof_suspend(struct device *dev)
+{
+	int ret;
+
+	if (sdev->dsp_state == SOF_DSP_D0I3 && d0i3_enable) {
```
I think it's really better to query the pm_runtime state (active or suspended) and only deal with D0ix when the pm_runtime state is 'active'.
You are duplicating work and information here.

keyonjie replied:
Thanks, let me change this and get back.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -533,6 +535,7 @@ static int sof_pcm_close(struct snd_pcm_substream *substream)
 	sdev->pcm_active--;
 	if (spcm->stream[substream->stream].avail_at_lp)
 		sdev->lp_pcm_active--;
+	schedule_work(&sdev->d0ix_work);
```
so what is missing in this series is something that will run after e.g each IPC.
This is not just about opening/closing streams, we need to keep the DSP active while we are doing IPCs.

keyonjie replied:
Yes, I planed to do that with IPC with registry part(need alignment with FW), it is in the to-do list also.

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -384,6 +384,7 @@ struct snd_sof_dev {
 	enum sof_dsp_state dsp_state;
 	u32 pcm_active;		/* Count of active PCMs */
 	u32 lp_pcm_active;	/* Count of active low power PCMs */
+	struct work_struct d0ix_work;
```
@keyonjie does it make sense to move this to struct sof_intel_dsp_desc? This is Intel DSP specific and I'm thinking might not be relevant to other architectures or other types of devices when MFD comes into picture

keyonjie replied:
I would say this is designed generically, there is platform specific d0i3 configuration(handled in platform d0ix_udpate callback), but the delayed work is based on PCM and is generic for all SOF drivers.

ranj063 commented on sound/soc/sof/pcm.c:
```diff
@@ -496,6 +496,8 @@ static int sof_pcm_open(struct snd_pcm_substream *substream)
 	if (spcm->stream[substream->stream].avail_at_lp)
 		sdev->lp_pcm_active++;
 
+	schedule_work(&sdev->d0ix_work);
```
@keyonjie do you think it makes more sense to check if the stream is a D0ix compatible stream and if not cancel the D0ix work? And then schedule it when it is closed?

keyonjie replied:
The D0ix wok is run on generic sof level, and each pcm open/close() calling may introduce the D0ix state refreshing, e.g.
D0i0--(a new d0ix compatible pcm being opened, no any other pcm opened)->D0i3
D0i0--(the last opened normal pcm being closed, and d0ix compatible opened pcm still opened)-->D0i3
D0i3--(normal pcm open)-->D0i0
D0i3--(the only one d0ix compatible opened pcm being closed)-->D0i0

So, to be simplified, we schedule the d0ix_work here at each .open()/.close(), do quick check the the work and return if there is no any update needed.

plbossart replied:
@keyonjie what if the user changes a capture volume control? Or accesses SRAM windows to get a trace/reg dump. 
You cannot make decisions to enter D0ix based on .open() only, you have to track all interactions with the DSP, including IPC and accesses to SRAM windows.

keyonjie replied:
@plbossart correct, for kcontrol update(IPC actually), share memory access and IPC sending, I plan to do something like this:
1. d0i3->d0i0 force,
2. do the specific task(IPC sending or memory windows access),
3. d0i0->d0i3 if suitable.

Comments about this?

Let me update my PR to reflect what I figure out about how to handle it.

keyonjie commented:
Hi @plbossart @ranj063 updated to new version.

With this version, I can even resume the system from s2idle via keyword voice.

Still have one issue left, the D0I3 IPCs is not sent via registers only, will change this next Tuesday.

plbossart commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,50 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+	if (!retry) {
+		dev_err(bus->dev, "CIP timeout after D0I3C updated!\n");
+		return -ETIMEDOUT;
+	}
```
might be worth having a helper to read the register, it's not very nice as is.

keyonjie replied:
OK, I can split it out, make something like hda_dsp_wait_d0i3c_done(u32 retry).

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -724,6 +724,13 @@ static const struct sof_topology_token pcm_tokens[] = {
+static const struct sof_topology_token stream_tokens[] = {
+	{SOF_TKN_STREAM_AVAIL_ON_LP,
+		SND_SOC_TPLG_TUPLE_TYPE_BOOL, get_token_u16,
+		offsetof(struct snd_sof_pcm_stream, avail_at_lp), 0},
```
wrong values? How does this even compile?

keyonjie replied:
Sorry it should be merged with the corresponding part with the later commit, updating it now.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -293,6 +293,7 @@ struct snd_sof_pcm_stream {
 	struct sof_ipc_stream_posn posn;
 	struct snd_pcm_substream *substream;
 	struct work_struct period_elapsed_work;
+	bool d0i3_compatible; /* pcm can run at low power like d0i3 */
```
is this single direction or both capture/playback? in the latter case it needs to be an array.

keyonjie replied:
it's single direction.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -726,9 +726,12 @@ static const struct sof_topology_token pcm_tokens[] = {
 /* PCM */
 static const struct sof_topology_token stream_tokens[] = {
-	{SOF_TKN_STREAM_AVAIL_ON_LP,
+	{SOF_TKN_STREAM_PLAYBACK_COMPATIBLE_D0I3,
```
squash this earlier

keyonjie replied:
exactly, thanks.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -357,6 +357,12 @@ struct snd_sof_dai {
+/* DSP D0ix sub-state */
+enum sof_d0_state {
+	SOF_DSP_D0I0 = 0, /* DSP default D0 state */
+	SOF_DSP_D0I3, /* DSP D0i3(low power) state*/
```
try to align comments

plbossart replied:
and these are all substates

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -357,6 +357,12 @@ struct snd_sof_dai {
 };
 
+/* DSP D0ix sub-state */
+enum sof_d0_state {
```
then call it substate.

keyonjie replied:
how about sof_d0_substate?

plbossart replied:
yes

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -372,6 +378,7 @@ struct snd_sof_dev {
 	struct snd_soc_component_driver plat_drv;
 
 	/* power status related */
+	enum sof_d0_state d0_state;
```
substate

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -385,6 +388,9 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 			ret);
 
+	/* DSP entered D3, reset default D0 state */
+	sdev->d0_state = SOF_DSP_D0I0;
```
Sorry @keyonjie this is WRONG. You cannot be in D3 and a substate be D0i0. That's just a bad idea that will raise issues with upstream reviews. D0ix is a substate of D0, therefore don't invent new concepts.

keyonjie replied:
OK, I just wanted to make sure it will be reset at suspend/resume, you are right, and here the resetting is superfluous also, let me remove it, thanks.

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -193,6 +193,15 @@ static inline int snd_sof_dsp_set_clk(struct snd_sof_dev *sdev, u32 freq)
+					  bool d0i3_enable)
+{
+	if (sof_ops(sdev)->d0ix_update)
+		return sof_ops(sdev)->d0ix_update(sdev, d0i3_enable);
```
I still believe you need to send a power state, not a boolean. This doesn't scale if we enable additional D0ix states.

keyonjie replied:
The only thing what this callback here do is set the D0I3C.D0i3 bit, it could be either "enable"(1, d0i3) or "disabled"(0, d0i0), no other choice.

plbossart replied:
this is a generic function, why would you make intel-specific (and generation-specific) assumptions?
Actually why does this even need to be exposed as a callback in an .ops? this is an operation you can do inside of the Intel driver after the IPC succeeds.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -425,3 +425,46 @@ int snd_sof_suspend(struct device *dev)
+ *  |         vv          v                 |        |substream
+ *  |    ---------------------        -------------- |opened
+ *  | No |only low power     |        |target state| |
+ *  +----|substream running  |        |    D0i0    | |Or force
```
what does 'running' mean? Use ALSA concepts please.

keyonjie replied:
let me use 'opened'.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -425,3 +425,46 @@ int snd_sof_suspend(struct device *dev)
+ *  |         ||d0i3 exit |                 |        |
+ *  | A normal||forcing   | A low power     |        |
+ *  |substream||task done,| substream   Yes |        |
+ *  | closed  ||return    | opened          |        |A normal
```
need more spaces so that the || is not close to words

plbossart replied:
why do we have "d0i3 exit forcing  task done, return" as a transition from D0i0 to D0i3, this looks just wrong or the wrong direction

keyonjie replied:
> need more spaces so that the || is not close to words

OK, let me try that.

keyonjie replied:
> why do we have "d0i3 exit forcing task done, return" as a transition from D0i0 to D0i3, this looks just wrong or the wrong direction

That is for something like sending IPC in D0i3: 1. force to exit from d0i3 -> 2. send IPC -> 3. re-enter d0i3, here it is the 3rd step.

plbossart replied:
ok, so my suggestion is to simplify the state machine by labeling the arcs, e.g. (1) (2) etc, and have test that describe what causes each transition. It's too hard to read otherwise and it's hard to extend.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -425,3 +425,46 @@ int snd_sof_suspend(struct device *dev)
+ *  (Delayed work)|             substream  |  | No   |D0I3 exit
+ *                v              closed    |  |      |for share mem
+ *       -------------------               |  |      |access
+ *       |                 |---------------+  |      |
```
move the stuff on the right further right, it's very difficult to read.

keyonjie replied:
OK, I was concerning about exceeding maximum 80.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -425,3 +425,46 @@ int snd_sof_suspend(struct device *dev)
+{
+	/* if only low power PCMs are running, target is D0i3 */
+	if (sdev->pcm_active == sdev->lp_pcm_active &&
+	    sdev->lp_pcm_active) {
```
so here you are not protected by a mutex, so how robust is this code?

keyonjie replied:
Um, good point, let me consider adding a mutex for it.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -425,3 +425,46 @@ int snd_sof_suspend(struct device *dev)
+/* Decide if we should try to enter D0i3 state or not */
+static int sof_d0ix_target_state(struct snd_sof_dev *sdev)
+{
+	/* if only low power PCMs are running, target is D0i3 */
```
don't use 'running'. use ALSA concepts, it's either OPENED, PREPARED, etc.

keyonjie replied:
OK.

plbossart commented on include/sound/sof/header.h:
```diff
@@ -74,6 +74,8 @@
 #define SOF_IPC_PM_CLK_REQ			SOF_CMD_TYPE(0x006)
 #define SOF_IPC_PM_CORE_ENABLE			SOF_CMD_TYPE(0x007)
+#define SOF_IPC_PM_D0I3_ENTER			SOF_CMD_TYPE(0x008)
+#define SOF_IPC_PM_D0I3_EXIT			SOF_CMD_TYPE(0x009)
```
this should be a single command with 'set_state' to scale to future extensions, and the CTX_SAVE/RESTORE should be deprecated to use this set_state command.

keyonjie replied:
As replied to the review summarization, I wanted to send D0I3_ENTER and D0I3_EXIT IPC without any payload. How about to keep as it is in this PR, and optimize them in the future, as the combining itself will require ABI change for S0<->S3 which could be another separated PR.

plbossart replied:
This is an ABI change,  we need to be careful about adding stuff and then remove it. it's a nightmare. let's do the right thing and define SET_STATE which for older firmware will be compatible with CTX_SAVE/RESTORE>

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -468,3 +468,35 @@ static int sof_d0ix_target_state(struct snd_sof_dev *sdev)
+	int ret;
+
+	if (sdev->d0_state == SOF_DSP_D0I3) {
+		dev_dbg(sdev->dev, "%s: already in D0i3.\n", __func__);
```
when would this happen? This looks like a bug to me.

keyonjie replied:
It should not happen, let me change to dev_err().

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -468,3 +468,35 @@ static int sof_d0ix_target_state(struct snd_sof_dev *sdev)
+		/* Todo: destroy unused pipelines/widgets in FW */
+
+		/* do platform specific d0i3 entering task */
+		snd_sof_dsp_d0ix_update(sdev, true);
```
need to test for error values

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -472,6 +475,8 @@ int snd_sof_runtime_idle(struct device *dev);
 int snd_sof_suspend(struct device *dev);
 void snd_sof_d0i3_entry_work(struct work_struct *work);
+void snd_sof_set_dsp_d0i0(struct snd_sof_dev *sdev, bool force, bool reload);
+void snd_sof_set_dsp_d0i3(struct snd_sof_dev *sdev);
```
set state, don't duplicate the work with multiple functions

keyonjie replied:
It is quite different for setting d0i0(no delayed work needed) and d0i3 in current implementation, let me figure out the runtime-PM solution first.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -39,6 +39,9 @@
 #define SND_SOF_SUSPEND_DELAY_MS	2000
 
+/* delay before scheduling D0i3 entry work */
+#define SND_SOF_D0I3_DELAY 5000
```
should this be part of topology?

keyonjie replied:
It could be configured via topology file also, but I don't see the need for that yet, how about keeping macro here(can act as fallback timeout delay if no topology definition specified in the future).

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -500,3 +500,58 @@ void snd_sof_d0i3_entry_work(struct work_struct *work)
+		dev_dbg(sdev->dev, "will schedule D0i3 work after %d ms\n",
+			SND_SOF_D0I3_DELAY);
+		schedule_delayed_work(&sdev->d0i3_entry_work,
+				      msecs_to_jiffies(SND_SOF_D0I3_DELAY));
```
if the helper is called, you'd need to cancel previous calls to 'schedule' so that the timer is reset to the last call to this helper (same as mark_last_busy for pm_runtime).

lyakh replied:
Wondering: this looks very similar to automatic delayed runtime-PM, can we maybe somehow reuse that API here? If we don't have a "free" device, to which we could meaningfully attach this RT-PM, maybe even add a new device for this.

keyonjie replied:
> Wondering: this looks very similar to automatic delayed runtime-PM, can we maybe somehow reuse that API here? If we don't have a "free" device, to which we could meaningfully attach this RT-PM, maybe even add a new device for this.

This is a fresh idea, let me consider it and do some investigation on creating new device for it. Thanks, @lyakh

keyonjie replied:
> if the helper is called, you'd need to cancel previous calls to 'schedule' so that the timer is reset to the last call to this helper (same as mark_last_busy for pm_runtime).

It looks not easy to handle this decently, as @lyakh suggested, let me consider reusing runtime-PM APIs.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -290,12 +290,20 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
+	 * D0I3_EXIT is exception.
+	 */
+	if (header != (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_D0I3_EXIT))
+		snd_sof_set_dsp_d0i0(sdev, true, true);
```
error handling?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -304,6 +312,10 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
 
+	/* go back to D0i3 if possible, D0I3_EXIT is exception. */
+	if (header != (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_D0I3_EXIT))
+		snd_sof_set_dsp_d0i3(sdev);
```
error handling?

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -153,6 +153,12 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
 {
+	if (sdev->d0_state == SOF_DSP_D0I3) {
+		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
+				  CNL_DSP_REG_HIPCIDR_BUSY | msg->header);
```
This looks wrong. You'd need to have a list of IPCs that use the registers and others that use the windows. we should not allow for generic use of the header for all messages, it's not quite right.

plbossart commented on sound/soc/sof/intel/hda-ipc.c:
```diff
@@ -56,6 +56,12 @@ static void hda_dsp_ipc_dsp_done(struct snd_sof_dev *sdev)
+		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
+				  CNL_DSP_REG_HIPCIDR_BUSY | msg->header);
+		return 0;
+	}
```
ditto, we should first fix the IPCs to use registers that allow all of them to use the registers. The header doesn't convey the information that's required in all cases.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -387,6 +387,7 @@ struct snd_sof_dev {
 	u32 pcm_active;		/* Count of active PCM substream */
 	u32 lp_pcm_active;	/* Count of active low power PCM substream */
 	struct delayed_work d0i3_entry_work;
+	bool kwd_at_s0i3; /* KWD run at S0I3 suspended */
```
no, this is too specific and shouldn't be in snd_sof_dev.
And it's not clear to me why this is needed at all.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -361,6 +365,12 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 	case SNDRV_PCM_TRIGGER_SUSPEND:
+		if (spcm->stream[substream->stream].d0i3_compatible) {
+			/* KWD keep working, don't handle DAPM PMD event */
+			sdev->kwd_at_s0i3 = true;
```
no, this is horrible. Don't introduce something that is so specific in a generic code. NAK on this one.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -260,15 +260,61 @@ static void sof_cache_debugfs(struct snd_sof_dev *sdev)
+ *          |          v                   |
+ *          |  -------------------         |
+ *   resume |  | kwd_at_s0i3 &&  |   Yes   |
+ *          |  | target_is_s0i3  |---------+
```
No, please don't mix D and S states this way. The state machine starts with S states which then impacts D states, you are doing things in the wrong order/direction.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -260,15 +260,61 @@ static void sof_cache_debugfs(struct snd_sof_dev *sdev)
 		return 0;
 
+	if (sdev->kwd_at_s0i3 &&
+	    acpi_target_system_state() == ACPI_STATE_S0) {
```
so what you are trying to say is that the system is trying to enter S0ix while the device is in D0ix. You absolutely don't need this 'kdw_at_s0i3' kludge.

keyonjie replied:
maybe the naming here is confusion, any suggestion?

plbossart replied:
no it's not a naming issue. it's the fact that you already have a substate, so use it!

keyonjie commented:
> @keyonjie there are really good ideas in this PR and really bad parts that need to be looked at. The 

Thanks for reviewing.

> main concern is that you 'invent' new states, and you really need to align with ALSA and PM ones. 

I have changed to only creating new D0 sub-states, so ALSA/PM ones are handling D0<->D3, and SOF driver will handle D0i0<->D0i3 one internally(transparent to ALSA/PM framework), make sense?

> Also you are trying to use the register-based IPC but that should be a different PR to be generic. 

OK, let me split them out. I added them here with [RFC], aimed to get feedback and alignment about the implementation itself first.

> And last you need to use a 'set_state' approach instead of being based on D0ix only. We want to deprecate the CTX_SAVE/RESTORE since it's really duplicating what SET_STATE could do

CTX_SAVE/RESTORE is for D3/D0 set at PM, they are using big IPC payload with struct sof_ipc_pm_ctx, while for D0I3 exit, we want to use 0 payload IPC. And yes, I admit that we can combined all those to a unified IPC message like SET_STATE with 'state' (could be D0, D0i3, D3) as the payload.

We discussed about this yesterday with FW team, @mmaka1 agreed with me that using separated D0I3_ENTER and D0I3_EXIT IPC without any payload would be better(and more friendly for IPC via registry) in this case.

lyakh commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,50 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+	}
+
+	if (enable)
+		reg = reg | SOF_HDA_VS_D0I3C_I3;
```
reg |= SOF_HDA_...

lyakh commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,50 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+	if (enable)
+		reg = reg | SOF_HDA_VS_D0I3C_I3;
+	else
+		reg = reg & (~SOF_HDA_VS_D0I3C_I3);
```
reg &= ~SOF_HDA_...

lyakh commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,50 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+	/* Write to D0I3C after Command-In-Progress bit is cleared */
+	while ((reg & SOF_HDA_VS_D0I3C_CIP) && --retry) {
+		usleep_range(10, 15);
+		reg = snd_hdac_chip_readb(bus, VS_D0I3C);
```
This pattern occurs so often - when you have to perform a calculation to verify a condition and then repeat an action while the condition holds, so you end up with
```
test = calculate_test();
while (must_loop(test)) {
	action(test);
	test = calculate_test();
}
```
that maybe we have to add an exception to kernel coding style guidelines to allow
```
while (must_loop(test = calculate_test()))
	action(test);
```
but for now we have to do the former...

keyonjie replied:
So what's your suggestion here @lyakh ?

lyakh replied:
@keyonjie write to LKML to suggest add this exception to the Coding Style :-)

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -2304,6 +2308,15 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
+	ret = sof_parse_tokens(scomp, &spcm->stream[stream], stream_tokens,
+			       ARRAY_SIZE(stream_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
```
doesn't checkpatch propose to just use `if (ret)`? Also below.

plbossart replied:
Agree with @lyakh, please run check patch.

keyonjie replied:
It didn't complain, I just manually run it again, no any warnings, anyway, let me change.

lyakh commented on sound/soc/sof/sof-priv.h:
```diff
@@ -371,6 +371,10 @@ struct snd_sof_dev {
 
+	/* power status related */
+	u32 pcm_active;		/* Count of active PCM substream */
+	u32 lp_pcm_active;	/* Count of active low power PCM substream */
```
no need for fixed-width, just use unsigned int.

keyonjie replied:
make sense.

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -500,3 +500,58 @@ void snd_sof_d0i3_entry_work(struct work_struct *work)
+	}
+
+	if (sof_d0ix_target_state(sdev) == SOF_DSP_D0I3 &&
+	    !force) {
```
cannot the above two lines be joined? That doesn't seem to exceed 80 characters.

keyonjie replied:
Yes, it can. It couldn't previously with longer naming. Let me change, thanks.

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -500,3 +500,58 @@ void snd_sof_d0i3_entry_work(struct work_struct *work)
+	cancel_delayed_work_sync(&sdev->d0i3_entry_work);
+
+	/* do platform specific d0i3 exit */
+	snd_sof_dsp_d0ix_update(sdev, false);
```
was it this call, that @plbossart had suggested to use a state, instead of a bool?

lyakh commented on sound/soc/sof/pcm.c:
```diff
@@ -531,9 +536,16 @@ static int sof_pcm_close(struct snd_pcm_substream *substream)
+		/* may be possible to enter d0i3 */
+		snd_sof_set_dsp_d0i3(sdev);
+	}
+
```
I actually would rather comment to the commit message of this patch - it looks very much like re-inventing runtime-PM to me, can we convert this to use the standard RT-PM APIs?

plbossart replied:
yes, it feels like re-invention but no it's not. pm_runtime is only about DO-D3. When the device is in D0i3 (which is platform-specific), pm_runtime considers the device as 'active'. So indeed we need to implement a custom usage/refcount and decision making. It's not necessarily elegant but it's not generic either so the ability to rely on a common framework for stuff that's hardware/firmware specific is limited.

keyonjie replied:
Hi @lyakh and @plbossart I think using RT-PM API is doable, it only define runtime_status as either suspended or active, not limit them to be D3/D0, we can make it suspended==d0i3 and active==d0i0.

I did a prototype code to have a quick try today, it works fine. What I did is:
1. Introduce a new platform device(child of pci device) with PM ops(runtime suspend/resume).
2. runtime get at pcm_open() and put at pcm_close() for normal substream only.
3. do d0i3 entry at runtime_suspend(), d0i3 exit at runtime_resume().

@plbossart if you can accept this solution, I will clean up and update it to the PR tomorrow.

lyakh replied:
@keyonjie I see @plbossart's point that RTPM is only about D0 / D3, but still doing what we want to do "from scratch" doesn't seem very pretty to me. At the very least I'd look at using kref instead of doing manual use-counting. But even better I'd talk to Linux PM gurus. Hint:: @rjwysocki is at Intel

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -260,15 +260,61 @@ static void sof_cache_debugfs(struct snd_sof_dev *sdev)
+			dev_warn(sdev->dev,
+				 "warning: failed to resuming trace from S0I3 %d\n",
+				 ret);
+		}
```
superfluous braces

keyonjie replied:
Thanks.

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -344,6 +391,14 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 
+	if (sdev->kwd_at_s0i3 &&
+	    acpi_target_system_state() == ACPI_STATE_S0) {
+		dev_dbg(sdev->dev, "KWD will work in S0i3, don't suspend adsp.\n");
```
capitalise "ADSP"

keyonjie replied:
OK.

keyonjie commented:
@plbossart @lyakh haven't addressed all comments yet and time has been used up, I will try implement with runtime-PM APIs tomorrow and get back after that.

plbossart commented:
>     main concern is that you 'invent' new states, and you really need to
>     align with ALSA and PM ones.
> 
> I have changed to only creating new D0 sub-states, so ALSA/PM ones are 
> handling D0<->D3, and SOF driver will handle D0i0<->D0i3 one 
> internally(transparent to ALSA/PM framework), make sense?

you have this notion of a 'running' device, which doesn't exist in 
ALSA-land. Also you redefined this kb_at_d0i3 thing when what you could 
do is an logical AND of all stream states to define the device state.

> 
>     Also you are trying to use the register-based IPC but that should be
>     a different PR to be generic.
> 
> OK, let me split them out. I added them here with [RFC], aimed to get 
> feedback and alignment about the implementation itself first.
> 
>     And last you need to use a 'set_state' approach instead of being
>     based on D0ix only. We want to deprecate the CTX_SAVE/RESTORE since
>     it's really duplicating what SET_STATE could do
> 
> CTX_SAVE/RESTORE is for D3/D0 set at PM, they are using big IPC payload 
> with struct struct sof_ipc_pm_ctx, while for D0I3 exit, we want to use 0 
> payload IPC. And yes, I admit that we can combined all those to a 
> unified category like SET_STATE with 'state' (could be D0, D0i3, D3) as 
> the payload.
> 
> We discussed about this yesterday with FW team, @mmaka1 
> <https://github.com/mmaka1> agreed with me that using separated 
> D0I3_ENTER and D0I3_EXIT IPC without any payload would be better(and 
> more friendly for IPC via registry) in this case.

I think we are confusing layers. You could have a 'set-state' IPC at the 
top level, and in the Intel drivers translate this as a bespoke 
register-based transfer with no payload. I really don't want a new 
intel-specific IPC visible for all platforms, it's not right, sorry.

keyonjie commented:
Hi @lyakh @plbossart I have changed to use runtime PM to a new platform device for this D0i0<->d0i3 transition management, the logic looks much simpler now. please help review and give your comments, thanks.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -62,6 +62,12 @@
 
 #define DMA_CHAN_INVALID	0xFFFFFFFF
 
+/* DSP D0ix sub-state */
```
change the commit message, it's now d0_substate

keyonjie replied:
OK, thanks.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -405,6 +405,9 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 
+	/* initialize default D0 sub-state */
+	sdev->d0_substate = SOF_DSP_D0I0;
+
```
I would do the initialization earlier, since the sof_probe_complete is where you typically enable pm_runtime

keyonjie replied:
makes sense.

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -193,6 +193,15 @@ static inline int snd_sof_dsp_set_clk(struct snd_sof_dev *sdev, u32 freq)
 	return 0;
 }
 
+static inline int snd_sof_dsp_set_state(struct snd_sof_dev *sdev,
```
set_power_state

keyonjie replied:
OK.

plbossart commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,52 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+	}
+}
+
+int hda_dsp_set_state(struct snd_sof_dev *sdev,
```
set_power_state

plbossart commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,52 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
 			HDA_DSP_REG_HIPCCTL_BUSY | HDA_DSP_REG_HIPCCTL_DONE, 0);
 }
 
+static void hda_dsp_wait_d0i3c_done(struct snd_sof_dev *sdev, int *retry)
```
this handles retries so should return an error if there is no change after all the retries

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -452,6 +456,9 @@ int hda_dsp_core_reset_power_down(struct snd_sof_dev *sdev,
 void hda_dsp_ipc_int_enable(struct snd_sof_dev *sdev);
 void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev);
 
+int hda_dsp_set_state(struct snd_sof_dev *sdev,
```
suggestion:
set_power_state(sdev, dstate, d0_substate).
That way in a future PR we can deprecate CTX_SAVE/RESTORE and use this set_power_state instead

keyonjie replied:
OK, thanks.

keyonjie replied:
Sorry, @plbossart you didn't get my point here. This helper is only aimed to set d0i3c register, not including sending IPC to FW, so the CTX_SAVE/RESTORE should not belong to here.

plbossart commented on include/sound/sof/header.h:
```diff
@@ -74,6 +74,7 @@
 #define SOF_IPC_PM_CLK_GET			SOF_CMD_TYPE(0x005)
 #define SOF_IPC_PM_CLK_REQ			SOF_CMD_TYPE(0x006)
 #define SOF_IPC_PM_CORE_ENABLE			SOF_CMD_TYPE(0x007)
+#define SOF_IPC_PM_SET_STATE			SOF_CMD_TYPE(0x008)
```
SET_POWER_STATE

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -422,3 +422,35 @@ int snd_sof_suspend(struct device *dev)
+	ret = snd_sof_dsp_set_state(sdev, d0_substate);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
```
remove d0i3, this this set_power_state ipc error

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -422,3 +422,35 @@ int snd_sof_suspend(struct device *dev)
+	ret = sof_send_pm_ipc(sdev, SOF_IPC_PM_SET_STATE | d0_substate);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
```
same here

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -468,6 +468,8 @@ int snd_sof_runtime_resume(struct device *dev);
 int snd_sof_runtime_idle(struct device *dev);
 int snd_sof_resume(struct device *dev);
 int snd_sof_suspend(struct device *dev);
+int snd_sof_set_dsp_state(struct snd_sof_dev *sdev,
```
please be careful with naming, you use set_state, set_dsp_state, but what you are doing is set_power_state()

keyonjie replied:
OK, let me change it to set_power_state

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -422,3 +422,35 @@ int snd_sof_suspend(struct device *dev)
+	}
+
+	/* sending set_state IPC */
+	ret = sof_send_pm_ipc(sdev, SOF_IPC_PM_SET_STATE | d0_substate);
```
and isn't it the other way around, you first send the IPC then set the D0i3C bit?

plbossart commented on sound/soc/sof/pm_d0ix.c:
```diff
@@ -0,0 +1,84 @@
+	.remove = pm_d0ix_remove,
+};
+
+module_platform_driver(sof_d0ix_driver);
```
Adding platform devices is not really something we want. We cannot control when the probe happens and in general it adds more issues.

I am inches away from stating 'NAK'

lyakh replied:
@keyonjie I'm also not too keen on adding a platform device just for the purpose of PM. It doesn't seem to be very logical to me. I'll comment with more explanations where @plbossart has explained why RTPM isn't the right approach here.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -486,10 +487,16 @@ static int sof_pcm_open(struct snd_pcm_substream *substream)
+	}
+
+	if (!spcm->stream[substream->stream].d0i3_compatible)
+		pm_runtime_get_sync(&plat_data->pdev_d0ix->dev);
```
no, this is officially horrible.

NAK NAK NAK

keyonjie commented:
> Sorry @keyonjie adding a new platform device just to manage power states is not something we even want to contemplate. been there, done that, and it creates a whole new set of problems. if you recall we already removed the platform devices between the PCI layer and SOF because we couldn't get pm_runtime to work reliably.

Hi @plbossart, yes, I recall the removal of the platform device(done by @ranj063 ?), but that was then. If you are concerned about adding a new platform device just to manage power states, I had the same feelings, that's why I asked your feedback about the idea above(at the point @lyakh commented).

But, if what you concern is it will create new problems for the runtime PM of the PCI device, I would say I don't worry about that too much here, the new device here(hereinafter referred to as "child B") is a child device of the PCI device, the child here is very similar to another child(the machine platform device, hereinafter referred to as "child A"), with only a minor difference that the d0i3 compatible stream will not hold the device. That means, this Child B is easier to enter "runtime suspended" state, comparing to Child A, and if no d0i3 compatible stream is available, the two children will be in same power state.

And, the pros with this solution is, we don't need to reimplement refcount(need lock or mutex) and delayed work(need cancel/flush the not finished work at closing or re-schedule) managements, which are actually error-prone to me.

@plbossart I am fine if you prefer my previous solution with d0i3 enter delayed work, but personally I tend to use this RT-PM device solution.

@lyakh @ranj063 @kv2019i @lgirdwood what's your opinion?

plbossart commented:
> > Sorry @keyonjie adding a new platform device just to manage power states is not something we even want to contemplate. been there, done that, and it creates a whole new set of problems. if you recall we already removed the platform devices between the PCI layer and SOF because we couldn't get pm_runtime to work reliably.
> 
> Hi @plbossart, yes, I recall the removal of the platform device(done by @ranj063 ?), but that was then. If you are concerned about adding a new platform device just to manage power states, I had the same feelings, that's why I asked your feedback about the idea above(at the point @lyakh commented).
> 
> But, if what you concern is it will create new problems for the runtime PM of the PCI device, I would say I don't worry about that too much here, the new device here(hereinafter referred to as "child B") is a child device of the PCI device, the child here is very similar to another child(the machine platform device, hereinafter referred to as "child A"), with only a minor difference that the d0i3 compatible stream will not hold the device. That means, this Child B is easier to enter "runtime suspended" state, comparing to Child A, and if no d0i3 compatible stream is available, the two children will be in same power state.
> 
> And, the pros with this solution is, we don't need to reimplement refcount(need lock or mutex) and delayed work(need cancel/flush the not finished work at closing or re-schedule) managements, which are actually error-prone to me.
> 
> @plbossart I am fine if you prefer my previous solution with d0i3 enter delayed work, but personally I tend to use this RT-PM device solution.
> 
> @lyakh @ranj063 @kv2019i @lgirdwood what's your opinion?

Once you create a child device, you have to assume that all decisions are then handled locally in the driver for the platform device.
That's not what you are doing, in the middle of the HDAudio code you have a pm_runtime_get_sync(childB), so you are NOT using a parent-child relationship, you are using the parent to change the state of the child.
That's not quite right.

I need to add that the parent-child model is that if the child wants to become active the parent will become active first, and conversely where the parent wants to suspend the child will be suspended first. Your solution adds another case that's moving away from the simple tried-and-tested concepts. When it's too good to be true it often is.

ranj063 commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,52 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+	int retry = 50;
+
+	/* Write to D0I3C after Command-In-Progress bit is cleared */
+	hda_dsp_wait_d0i3c_done(sdev, &retry);
```
@keyonjie why do we need to pass this as a pointer? Why not just initialize retry in the other function and check the return value for hda_dsp_wait_d0i3c_done() here. It seems a bit of an overkill to pass a pointer for no good reason

ranj063 replied:
@keyonjie do we need to add a check to see if the previous substate is already the requested state and return in tthat case?

lyakh replied:
agree, remove the parameter and return -ETIMEDOUT if retries exhausted, maybe also use a macro for the retry count

ranj063 commented on sound/soc/sof/intel/apl.c:
```diff
@@ -97,6 +97,7 @@ const struct snd_sof_dsp_ops sof_apl_ops = {
 	.runtime_resume		= hda_dsp_runtime_resume,
 	.runtime_idle		= hda_dsp_runtime_idle,
 	.set_hw_params_upon_resume = hda_dsp_set_hw_params_upon_resume,
+	.set_state		= hda_dsp_set_state,
```
@keyonjie incorrect commit title and message. Please align with cnl commit.

keyonjie replied:
Thanks.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2335,6 +2345,15 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
 	if (!spcm->pcm.capture)
 		return ret;
 
+	ret = sof_parse_tokens(scomp, &spcm->stream[stream], stream_tokens,
```
@keyonjie is this second call necessary? The first call to parse stream_tokens will have already parsed both the PLAYBACK and CAPTURE tokens if they were available.

keyonjie replied:
@ranj063 yes, it is necessary, we only parse one token each time, here we are using spcm->stream[stream] with direction stream.

ranj063 commented:
> @lyakh @ranj063 @kv2019i @lgirdwood what's your opinion?

@keyonjie I must say the addition of the platform device completely threw me off in this PR. It is not quite as much the addition of the platform device that looks weird as much as the how you are handling the usage counts. Like Pierre mentioned we cannot mess with refcounts for a child by calling the get/put methods from the parent.

Also the other thing to keep in mind is that if you add this platform device, it will be included in the system sleep PM state transitions as well. It almost sounds incorrect when you going from D0->D3, where the platform devices suspend() method will get called and you'll be in fact transitioning the DSP to D0i3? It sounds too fishy to me

ranj063 commented:
@keyonjie the other thing is that I am working on the introduction of MFD in the driver and with that the use of a platform device for the DSP driver is inevitable. I should have the RFC patches in a couple of days.

keyonjie commented:
> Once you create a child device, you have to assume that all decisions are then handled locally in the driver for the platform device.
> That's not what you are doing, in the middle of the HDAudio code you have a pm_runtime_get_sync(childB), so you are NOT using a parent-child relationship, you are using the parent to change the state of the child.
> That's not quite right.

w.r.t. your concern about changing state of the child in parent(or in another child), I can fix this, I have even crazier thought that making ChildB as a child of the ChildA. If you will consider this platform device solution basically, let me update my PR to reflect that.

> 
> I need to add that the parent-child model is that if the child wants to become active the parent will become active first, and conversely where the parent wants to suspend the child will be suspended 

Yes, that's true, and it is exactly matched in our scenario here?

> first. Your solution adds another case that's moving away from the simple tried-and-tested concepts. When it's too good to be true it often is.

Sorry, don't quite catch here, are you talking about the count in .open()/.close()?

keyonjie commented:
> > @lyakh @ranj063 @kv2019i @lgirdwood what's your opinion?
> 
> @keyonjie I must say the addition of the platform device completely threw me off in this PR. It is not quite as much the addition of the platform device that looks weird as much as the how you are handling the usage counts. Like Pierre mentioned we cannot mess with refcounts for a child by calling the get/put methods from the parent.
> 
> Also the other thing to keep in mind is that if you add this platform device, it will be included in the system sleep PM state transitions as well. It almost sounds incorrect when you going from D0->D3, where the platform devices suspend() method will get called and you'll be in fact transitioning the DSP to D0i3? It sounds too fishy to me

@ranj063 if you go through the driver for the new platform device, you will find that there is no .suspend()/.resume() callbacks in its PM ops, we only use its runtime-PM logic.

ranj063 commented on sound/soc/sof/pm_d0ix.c:
```diff
@@ -0,0 +1,84 @@
+static int pm_d0ix_remove(struct platform_device *pdev)
+{
+	return 0;
+}
```
@keyonjie is remove() really needed if you dont do anything here?

ranj063 commented:
> @ranj063 if you go through the driver for the new platform device, you will find that there is no .suspend()/.resume() callbacks in its PM ops, we only use its runtime-PM logic.

@keyonjie thanks for pointing this out. I missed noticing this. Anyway, this makes it even more questionable. When you wake up from system suspend, the intention is that all devices are resumed() whether you have a system resume callback or not. And eventually after the autosuspend delay, the runtime_suspend callback will be invoked which means you'll be setting the DSp state to D0i3? is that correct?

plbossart commented:
> > Once you create a child device, you have to assume that all decisions are then handled locally in the driver for the platform device.
> > That's not what you are doing, in the middle of the HDAudio code you have a pm_runtime_get_sync(childB), so you are NOT using a parent-child relationship, you are using the parent to change the state of the child.
> > That's not quite right.
> 
> w.r.t. your concern about changing state of the child in parent(or in another child), I can fix this, I have even crazier thought that making ChildB as a child of the ChildA. If you will consider this platform device solution basically, let me update my PR to reflect that.

there is no childA btw, we've removed all platform devices except for the machine driver.

> > I need to add that the parent-child model is that if the child wants to become active the parent will become active first, and conversely where the parent wants to suspend the child will be suspended
> 
> Yes, that's true, and it is exactly matched in our scenario here?

no, you are adding a direct interaction between parent and child with this:

if (!spcm->stream[substream->stream].d0i3_compatible)
		pm_runtime_get_sync(&plat_data->pdev_d0ix->dev);

That is the part that breaks the hierarchy. We should not go there, it's error prone and not recommended in the first place.
> 
> > first. Your solution adds another case that's moving away from the simple tried-and-tested concepts. When it's too good to be true it often is.
> 
> Sorry, don't quite catch here, are you talking about the count in .open()/.close()?

No, I am talking about the design with the interaction between parent and child, and the fact that child will initiate an IPC that is handled at the parent level (in the PCI driver). It's just a recipe for bad things to happen.

lyakh commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -282,6 +282,52 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
+
+	if (!retry) {
+		dev_err(bus->dev, "CIP timeout before update D0I3C!\n");
+		return -ETIMEDOUT;
```
so here you'll just propagate the error back to the caller

keyonjie commented:
> > @ranj063 if you go through the driver for the new platform device, you will find that there is no .suspend()/.resume() callbacks in its PM ops, we only use its runtime-PM logic.
> 
> @keyonjie thanks for pointing this out. I missed noticing this. Anyway, this makes it even more questionable. When you wake up from system suspend, the intention is that all devices are resumed() whether you have a system resume callback or not. And eventually after the autosuspend delay, the runtime_suspend callback will be invoked which means you'll be setting the DSp state to D0i3? is that correct?

Yes, that's correct, if no normal substream is opened after the system resumed.

keyonjie commented:
OK, I realized that the conversation is getting exploded and difficult to follow. Let me submit some aligned preliminaries firstly and make the opens to be smaller.

Here it is: https://github.com/thesofproject/linux/pull/1190

plbossart commented:
See below quote from Greg KH:

````text
Don't mess with a platform device unless you really have no other possible choice.  And even then, don't do it and try to do something else.  Platform devices are really abused, don't perpetuate it
````

plbossart commented:
@keyonjie this PR looks abandoned, please close or maintain it, but don't leave it in zombie state.

keyonjie commented:
@plbossart sorry, just focused on issues fixing these days, I didn't close this, as I thought the discussion here is valuable for the D0i3 support on S0.

But yes, I can do that with it closed also, closing it now.
